### PR TITLE
[FLINK-4793] [types] Improve lambda constructor reference handling

### PR DESCRIPTION
**Pull request by twalthr**
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [x] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [x] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [x] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

This PR further improves the handling of Java 8 lambdas. It improves the handling of constructor references. The following is now possible:

`env.fromElements("xx", "xx").map(Trade::new).map(Trade::toString).print();`

Unfortunately, I had to add the ASM dependency to `flink-core`. I hope this is not a problem.

### CODE REVIEW

File: flink-core/src/main/java/org/apache/flink/api/java/typeutils/TypeExtractionUtils.java [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+				for (Method method : methods) {
+					if(method.getName().equals(methodName) && getMethodDescriptor(method).equals(methodSig)) {
+						return new LambdaExecutable(method);
+					}
+				}
+			}
+			throw new Exception("No lambda method found.");
+		}
+		catch (Exception e) {
+			throw new RuntimeException("Could not extract lambda method out of function: " + e.getClass().getSimpleName() + " - " + e.getMessage(), e);
```
Comment (StephanEwen) [inline flink-core/src/main/java/org/apache/flink/api/java/typeutils/TypeExtractionUtils.java]: RuntimeException is usually used for programming errors (wrong arguments, null divisions, ...).
Something that is expected to fail sometimes beyond the programmers control (I/O, code analysis, etc) should use checked exceptions.

Comment (StephanEwen) [inline flink-core/src/main/java/org/apache/flink/api/java/typeutils/TypeExtractionUtils.java]: RuntimeException is usually used for programming errors (wrong arguments, null divisions, ...).
Something that is expected to fail sometimes beyond the programmers control (I/O, code analysis, etc) should use checked exceptions.

Comment (twalthr) [inline flink-core/src/main/java/org/apache/flink/api/java/typeutils/TypeExtractionUtils.java]: I will use an `IllegalStateException`.

Comment (StephanEwen) [inline flink-core/src/main/java/org/apache/flink/api/java/typeutils/TypeExtractionUtils.java]: That is another unchecked exception ;-)

Comment (twalthr) [inline flink-core/src/main/java/org/apache/flink/api/java/typeutils/TypeExtractionUtils.java]: Sorry, you are right. Ok then I will introduce a new one.

**Pull request closed without merging**